// devextreme/esm/renovation/ui/scheduler/appointment/tooltip/appointment_list.js
/**
 * DevExtreme (esm/renovation/ui/scheduler/appointment/tooltip/appointment_list.js)
 * Version: 23.2.5
 * Build date: Mon Mar 11 2024
 *
 * Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/esm/objectWithoutPropertiesLoose";
var _excluded = ["appointments"];
import {
    createVNode,
    createComponentVNode
} from "inferno";
import {
    BaseInfernoComponent
} from "@devextreme/runtime/inferno";
import {
    TooltipItemLayout
} from "./item_layout";
export var viewFunction = viewModel => createVNode(1, "div", null, viewModel.props.appointments.map((item, index) => createComponentVNode(2, TooltipItemLayout, {
    item: item,
    index: index
}, item.key)), 0);
export var AppointmentListProps = {};
export class AppointmentList extends BaseInfernoComponent {
    constructor(props) {
        super(props);
        this.state = {}
    }
    get restAttributes() {
        var _this$props = this.props,
            restProps = _objectWithoutPropertiesLoose(_this$props, _excluded);
        return restProps
    }
    render() {
        var props = this.props;
        return viewFunction({
            props: _extends({}, props),
            restAttributes: this.restAttributes
        })
    }
}
AppointmentList.defaultProps = AppointmentListProps;
