// devextreme/esm/renovation/ui/scheduler/appointment/tooltip/appointment_tooltip.js
/**
 * DevExtreme (esm/renovation/ui/scheduler/appointment/tooltip/appointment_tooltip.js)
 * Version: 23.2.5
 * Build date: Mon Mar 11 2024
 *
 * Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/esm/objectWithoutPropertiesLoose";
var _excluded = ["dataList", "onVisibleChange", "target", "visible"];
import {
    createComponentVNode
} from "inferno";
import {
    BaseInfernoComponent
} from "@devextreme/runtime/inferno";
import {
    Tooltip
} from "../../../overlays/tooltip";
import {
    AppointmentList
} from "./appointment_list";
var wrapperAttr = {
    class: "dx-scheduler-appointment-tooltip-wrapper"
};
export var viewFunction = _ref => {
    var {
        props: {
            dataList: dataList,
            onVisibleChange: onVisibleChange,
            target: target,
            visible: visible
        }
    } = _ref;
    return createComponentVNode(2, Tooltip, {
        focusStateEnabled: false,
        hideOnOutsideClick: true,
        visible: visible,
        visibleChange: onVisibleChange,
        target: target,
        wrapperAttr: wrapperAttr,
        children: createComponentVNode(2, AppointmentList, {
            appointments: dataList
        })
    })
};
export var AppointmentTooltipProps = {};
import {
    convertRulesToOptions
} from "../../../../../core/options/utils";
export class AppointmentTooltip extends BaseInfernoComponent {
    constructor(props) {
        super(props);
        this.state = {}
    }
    get restAttributes() {
        var _this$props = this.props,
            restProps = _objectWithoutPropertiesLoose(_this$props, _excluded);
        return restProps
    }
    render() {
        var props = this.props;
        return viewFunction({
            props: _extends({}, props),
            restAttributes: this.restAttributes
        })
    }
}
AppointmentTooltip.defaultProps = AppointmentTooltipProps;
var __defaultOptionRules = [];
export function defaultOptions(rule) {
    __defaultOptionRules.push(rule);
    AppointmentTooltip.defaultProps = Object.create(Object.prototype, _extends(Object.getOwnPropertyDescriptors(AppointmentTooltip.defaultProps), Object.getOwnPropertyDescriptors(convertRulesToOptions(__defaultOptionRules))))
}
