// devextreme/esm/renovation/ui/scheduler/appointment/tooltip/delete_button.js
/**
 * DevExtreme (esm/renovation/ui/scheduler/appointment/tooltip/delete_button.js)
 * Version: 23.2.5
 * Build date: Mon Mar 11 2024
 *
 * Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/esm/objectWithoutPropertiesLoose";
var _excluded = ["onClick"];
import {
    createVNode,
    createComponentVNode
} from "inferno";
import {
    BaseInfernoComponent
} from "@devextreme/runtime/inferno";
import {
    Button
} from "../../../button";
export var viewFunction = () => createVNode(1, "div", "dx-tooltip-appointment-item-delete-button-container", createComponentVNode(2, Button, {
    className: "dx-tooltip-appointment-item-delete-button",
    icon: "trash",
    stylingMode: "text"
}), 2);
export var DeleteButtonProps = {};
export class DeleteButton extends BaseInfernoComponent {
    constructor(props) {
        super(props);
        this.state = {}
    }
    get restAttributes() {
        var _this$props = this.props,
            restProps = _objectWithoutPropertiesLoose(_this$props, _excluded);
        return restProps
    }
    render() {
        this.props;
        return viewFunction()
    }
}
DeleteButton.defaultProps = DeleteButtonProps;
