// devextreme/esm/renovation/ui/scheduler/appointment/tooltip/item_content.js
/**
 * DevExtreme (esm/renovation/ui/scheduler/appointment/tooltip/item_content.js)
 * Version: 23.2.5
 * Build date: Mon Mar 11 2024
 *
 * Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/esm/objectWithoutPropertiesLoose";
var _excluded = ["formattedDate", "text"];
import {
    createVNode
} from "inferno";
import {
    BaseInfernoComponent
} from "@devextreme/runtime/inferno";
export var viewFunction = viewModel => createVNode(1, "div", "dx-tooltip-appointment-item-content", [createVNode(1, "div", "dx-tooltip-appointment-item-content-subject", viewModel.props.text, 0), createVNode(1, "div", "dx-tooltip-appointment-item-content-date", viewModel.props.formattedDate, 0)], 4);
export var TooltipItemContentProps = {};
export class TooltipItemContent extends BaseInfernoComponent {
    constructor(props) {
        super(props);
        this.state = {}
    }
    get restAttributes() {
        var _this$props = this.props,
            restProps = _objectWithoutPropertiesLoose(_this$props, _excluded);
        return restProps
    }
    render() {
        var props = this.props;
        return viewFunction({
            props: _extends({}, props),
            restAttributes: this.restAttributes
        })
    }
}
TooltipItemContent.defaultProps = TooltipItemContentProps;
