// devextreme/esm/renovation/ui/scheduler/appointment/tooltip/item_layout.js
/**
 * DevExtreme (esm/renovation/ui/scheduler/appointment/tooltip/item_layout.js)
 * Version: 23.2.5
 * Build date: Mon Mar 11 2024
 *
 * Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/esm/objectWithoutPropertiesLoose";
var _excluded = ["index", "item"];
import {
    createVNode,
    createComponentVNode
} from "inferno";
import {
    BaseInfernoComponent
} from "@devextreme/runtime/inferno";
import {
    Marker
} from "./marker";
import {
    TooltipItemContent
} from "./item_content";
import {
    DeleteButton
} from "./delete_button";
export var viewFunction = viewModel => createVNode(1, "div", "dx-tooltip-appointment-item", [createComponentVNode(2, Marker), createComponentVNode(2, TooltipItemContent, {
    text: viewModel.text,
    formattedDate: viewModel.dateText
}), createComponentVNode(2, DeleteButton)], 4);
export var TooltipItemLayoutProps = {
    index: 0
};
export class TooltipItemLayout extends BaseInfernoComponent {
    constructor(props) {
        super(props);
        this.state = {}
    }
    get text() {
        return this.props.item.appointment.text
    }
    get dateText() {
        return this.props.item.info.dateText
    }
    get restAttributes() {
        var _this$props = this.props,
            restProps = _objectWithoutPropertiesLoose(_this$props, _excluded);
        return restProps
    }
    render() {
        var props = this.props;
        return viewFunction({
            props: _extends({}, props),
            text: this.text,
            dateText: this.dateText,
            restAttributes: this.restAttributes
        })
    }
}
TooltipItemLayout.defaultProps = TooltipItemLayoutProps;
