// devextreme/cjs/ui/range_slider.js
/**
 * DevExtreme (cjs/ui/range_slider.js)
 * Version: 23.2.5
 * Build date: Mon Mar 11 2024
 *
 * Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
exports.default = void 0;
var _size = require("../core/utils/size");
var _renderer = _interopRequireDefault(require("../core/renderer"));
var _events_engine = _interopRequireDefault(require("../events/core/events_engine"));
var _slider = _interopRequireDefault(require("./slider"));
var _ui = _interopRequireDefault(require("./slider/ui.slider_handle"));
var _component_registrator = _interopRequireDefault(require("../core/component_registrator"));
var _extend = require("../core/utils/extend");
var _common = require("../core/utils/common");
var _index = require("../events/utils/index");
var _message = _interopRequireDefault(require("../localization/message"));

function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    }
}
const RANGE_SLIDER_CLASS = "dx-rangeslider";
const RANGE_SLIDER_START_HANDLE_CLASS = "dx-rangeslider-start-handle";
const RANGE_SLIDER_END_HANDLE_CLASS = "dx-rangeslider-end-handle";
const RangeSlider = _slider.default.inherit({
    _supportedKeys: function() {
        const isRTL = this.option("rtlEnabled");
        const that = this;
        const _changeHandle = function(e, capturedHandle) {
            if (that.option("start") === that.option("end")) {
                that._capturedHandle = capturedHandle;
                e.target = that._capturedHandle;
                _events_engine.default.trigger(that._capturedHandle, "focus")
            }
        };
        const _setHandleValue = function(e, step, sign) {
            const isStart = (0, _renderer.default)(e.target).hasClass("dx-rangeslider-start-handle");
            const valueOption = isStart ? "start" : "end";
            let val = that.option(valueOption);
            step = that._valueStep(step);
            val += sign * (isRTL ? -step : step);
            that.option(valueOption, val)
        };
        const moveHandleRight = function(e, step) {
            _changeHandle(e, isRTL ? that._$handleStart : that._$handleEnd);
            _setHandleValue(e, step, 1)
        };
        const moveHandleLeft = function(e, step) {
            _changeHandle(e, isRTL ? that._$handleEnd : that._$handleStart);
            _setHandleValue(e, step, -1)
        };
        return (0, _extend.extend)(this.callBase(), {
            leftArrow: function(e) {
                this._processKeyboardEvent(e);
                moveHandleLeft(e, this.option("step"))
            },
            rightArrow: function(e) {
                this._processKeyboardEvent(e);
                moveHandleRight(e, this.option("step"))
            },
            pageUp: function(e) {
                this._processKeyboardEvent(e);
                moveHandleRight(e, this.option("step") * this.option("keyStep"))
            },
            pageDown: function(e) {
                this._processKeyboardEvent(e);
                moveHandleLeft(e, this.option("step") * this.option("keyStep"))
            },
            home: function(e) {
                this._processKeyboardEvent(e);
                const isStart = (0, _renderer.default)(e.target).hasClass("dx-rangeslider-start-handle");
                const valueOption = isStart ? "start" : "end";
                const startOption = isStart ? "min" : "start";
                const val = this.option(startOption);
                this.option(valueOption, val)
            },
            end: function(e) {
                this._processKeyboardEvent(e);
                const isStart = (0, _renderer.default)(e.target).hasClass("dx-rangeslider-start-handle");
                const valueOption = isStart ? "start" : "end";
                const endOption = isStart ? "end" : "max";
                const val = this.option(endOption);
                this.option(valueOption, val)
            }
        })
    },
    _getDefaultOptions: function() {
        return (0, _extend.extend)(this.callBase(), {
            start: 40,
            end: 60,
            value: [40, 60],
            startName: "",
            endName: ""
        })
    },
    _renderSubmitElement: function() {
        const $element = this.$element();
        this._$submitStartElement = (0, _renderer.default)("<input>").attr("type", "hidden").attr("name", this.option("startName")).appendTo($element);
        this._$submitEndElement = (0, _renderer.default)("<input>").attr("type", "hidden").attr("name", this.option("endName")).appendTo($element)
    },
    _initOptions: function(options) {
        this.callBase(options);
        const initialValue = this.initialOption("value");
        const value = this.option("value");
        if (value[0] === initialValue[0] && value[1] === initialValue[1]) {
            this.option("value", [this.option("start"), this.option("end")])
        } else {
            this.option({
                start: value[0],
                end: value[1]
            })
        }
    },
    _initMarkup: function() {
        this.$element().addClass("dx-rangeslider");
        this.callBase()
    },
    _renderContentImpl: function() {
        this._callHandlerMethod("repaint");
        this.callBase()
    },
    _renderHandle: function() {
        this._$handleStart = this._renderHandleImpl(this.option("start"), this._$handleStart).addClass("dx-rangeslider-start-handle");
        this._$handleEnd = this._renderHandleImpl(this.option("end"), this._$handleEnd).addClass("dx-rangeslider-end-handle");
        this._updateHandleAriaLabels()
    },
    _startHandler: function(args) {
        const e = args.event;
        const $range = this._$range;
        const rangeWidth = (0, _size.getWidth)($range);
        const eventOffsetX = (0, _index.eventData)(e).x - this._$bar.offset().left;
        const startHandleX = $range.position().left;
        const endHandleX = $range.position().left + rangeWidth;
        const rtlEnabled = this.option("rtlEnabled");
        const startHandleIsClosest = (rtlEnabled ? -1 : 1) * ((startHandleX + endHandleX) / 2 - eventOffsetX) > 0;
        this._capturedHandle = startHandleIsClosest ? this._$handleStart : this._$handleEnd;
        this.callBase(args)
    },
    _updateHandleAriaLabels: function() {
        this.setAria("label", _message.default.getFormatter("dxRangeSlider-ariaFrom")(this.option("dxRangeSlider-ariaFrom")), this._$handleStart);
        this.setAria("label", _message.default.getFormatter("dxRangeSlider-ariaTill")(this.option("dxRangeSlider-ariaTill")), this._$handleEnd)
    },
    _activeHandle: function() {
        return this._capturedHandle
    },
    _updateHandlePosition: function(e) {
        const rtlEnabled = this.option("rtlEnabled");
        const offsetDirection = rtlEnabled ? -1 : 1;
        const max = this.option("max");
        const min = this.option("min");
        let newRatio = this._startOffset + offsetDirection * e.event.offset / this._swipePixelRatio();
        newRatio = newRatio.toPrecision(12);
        const newValue = newRatio * (max - min) + min;
        this._updateSelectedRangePosition(newRatio, newRatio);
        _ui.default.getInstance(this._activeHandle()).fitTooltipPosition;
        this._changeValueOnSwipe(newRatio);
        const [startValue, endValue] = this._getActualValue();
        let $nextHandle;
        if (startValue === endValue) {
            if (newValue < startValue) {
                $nextHandle = this._$handleStart
            } else {
                $nextHandle = this._$handleEnd
            }
            _events_engine.default.trigger($nextHandle, "focus");
            if ($nextHandle && $nextHandle !== this._capturedHandle) {
                this._updateSelectedRangePosition((startValue - min) / (max - min), (endValue - min) / (max - min));
                this._toggleActiveState(this._activeHandle(), false);
                this._toggleActiveState($nextHandle, true);
                this._capturedHandle = $nextHandle
            }
            this._updateSelectedRangePosition(newRatio, newRatio);
            this._changeValueOnSwipe(newRatio)
        }
    },
    _updateSelectedRangePosition: function(leftRatio, rightRatio) {
        const rtlEnabled = this.option("rtlEnabled");
        const moveRight = this._capturedHandle === this._$handleStart && rtlEnabled || this._capturedHandle === this._$handleEnd && !rtlEnabled;
        const prop = moveRight ? "right" : "left";
        if (rtlEnabled ^ moveRight) {
            this._$range.css(prop, 100 - 100 * rightRatio + "%")
        } else {
            this._$range.css(prop, 100 * leftRatio + "%")
        }
    },
    _setValueOnSwipe: function(value) {
        const option = this._capturedHandle === this._$handleStart ? "start" : "end";
        let [start, end] = this._getActualValue();
        const max = this.option("max");
        const min = this.option("min");
        start = Math.min(Math.max(start, min), max);
        end = Math.min(Math.max(end, min), max);
        if ("start" === option) {
            start = value > end ? end : value
        } else {
            end = value < start ? start : value
        }
        if ("onHandleMove" === this.option("valueChangeMode")) {
            this.option("value", [start, end])
        } else {
            this._actualValue = [start, end];
            this._renderValue()
        }
    },
    _renderValue: function() {
        let [valStart, valEnd] = this._getActualValue();
        const min = this.option("min");
        const max = this.option("max");
        const rtlEnabled = this.option("rtlEnabled");
        valStart = Math.max(min, Math.min(valStart, max));
        valEnd = Math.max(valStart, Math.min(valEnd, max));
        if ("onHandleMove" === this.option("valueChangeMode")) {
            this._setOptionWithoutOptionChange("start", valStart);
            this._setOptionWithoutOptionChange("end", valEnd);
            this._setOptionWithoutOptionChange("value", [valStart, valEnd])
        }
        this._$submitStartElement.val((0, _common.applyServerDecimalSeparator)(valStart));
        this._$submitEndElement.val((0, _common.applyServerDecimalSeparator)(valEnd));
        const ratio1 = max === min ? 0 : (valStart - min) / (max - min);
        const ratio2 = max === min ? 0 : (valEnd - min) / (max - min);
        const startOffset = parseFloat((100 * ratio1).toPrecision(12)) + "%";
        const endOffset = parseFloat((100 * (1 - ratio2)).toPrecision(12)) + "%";
        !this._needPreventAnimation && this._setRangeStyles({
            right: rtlEnabled ? startOffset : endOffset,
            left: rtlEnabled ? endOffset : startOffset
        });
        _ui.default.getInstance(this._$handleStart).option("value", valStart);
        _ui.default.getInstance(this._$handleEnd).option("value", valEnd)
    },
    _callHandlerMethod: function(name, args) {
        _ui.default.getInstance(this._$handleStart)[name](args);
        _ui.default.getInstance(this._$handleEnd)[name](args)
    },
    _setValueOption: function() {
        const start = this.option("start");
        const end = this.option("end");
        this.option("value", [start, end])
    },
    _rangesAreEqual: (firstRange, secondRange) => firstRange[0] === secondRange[0] && firstRange[1] === secondRange[1],
    _optionChanged: function(args) {
        switch (args.name) {
            case "value": {
                if (this._rangesAreEqual(args.value, args.previousValue)) {
                    break
                }
                this._setOptionWithoutOptionChange("start", args.value[0]);
                this._setOptionWithoutOptionChange("end", args.value[1]);
                this._renderValue();
                const start = this.option("start");
                const end = this.option("end");
                const isDirty = !this._rangesAreEqual(this._initialValue, args.value);
                this.option("isDirty", isDirty);
                this._createActionByOption("onValueChanged", {
                    excludeValidators: ["disabled", "readOnly"]
                })({
                    start: start,
                    end: end,
                    value: [start, end],
                    event: this._valueChangeEventInstance,
                    previousValue: args.previousValue
                });
                this.validationRequest.fire({
                    value: [start, end],
                    editor: this
                });
                this._saveValueChangeEvent(void 0);
                break
            }
            case "start":
            case "end":
                this._setValueOption();
                break;
            case "startName":
                this._$submitStartElement.attr("name", args.value);
                break;
            case "endName":
                this._$submitEndElement.attr("name", args.value);
                break;
            case "name":
                break;
            default:
                this.callBase(args)
        }
    }
});
(0, _component_registrator.default)("dxRangeSlider", RangeSlider);
var _default = RangeSlider;
exports.default = _default;
module.exports = exports.default;
module.exports.default = exports.default;
