// devextreme/bundles/__internal/scheduler/m_constants.js
/**
 * DevExtreme (bundles/__internal/scheduler/m_constants.js)
 * Version: 23.2.5
 * Build date: Mon Mar 11 2024
 *
 * Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.VIEWS = exports.LIST_ITEM_DATA_KEY = exports.LIST_ITEM_CLASS = exports.HORIZONTAL_GROUP_ORIENTATION = exports.APPOINTMENT_SETTINGS_KEY = void 0;
const LIST_ITEM_DATA_KEY = "dxListItemData";
exports.LIST_ITEM_DATA_KEY = "dxListItemData";
const LIST_ITEM_CLASS = "dx-list-item";
exports.LIST_ITEM_CLASS = "dx-list-item";
const APPOINTMENT_SETTINGS_KEY = "dxAppointmentSettings";
exports.APPOINTMENT_SETTINGS_KEY = "dxAppointmentSettings";
const HORIZONTAL_GROUP_ORIENTATION = "horizontal";
exports.HORIZONTAL_GROUP_ORIENTATION = "horizontal";
const VIEWS = {
    DAY: "day",
    WEEK: "week",
    WORK_WEEK: "workWeek",
    MONTH: "month",
    TIMELINE_DAY: "timelineDay",
    TIMELINE_WEEK: "timelineWeek",
    TIMELINE_WORK_WEEK: "timelineWorkWeek",
    TIMELINE_MONTH: "timelineMonth",
    AGENDA: "agenda"
};
exports.VIEWS = VIEWS;
