// devextreme/cjs/__internal/grids/pivot_grid/field_chooser/const.js
/**
 * DevExtreme (cjs/__internal/grids/pivot_grid/field_chooser/const.js)
 * Version: 23.2.5
 * Build date: Mon Mar 11 2024
 *
 * Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.SORT_ORDER = exports.SORTABLE_CONST = exports.ICONS = exports.CLASSES = exports.ATTRIBUTES = void 0;
const ATTRIBUTES = {
    treeViewItem: "tree-view-item",
    allowScrolling: "allow-scrolling",
    itemGroup: "item-group"
};
exports.ATTRIBUTES = ATTRIBUTES;
const CLASSES = {
    area: {
        self: "dx-area",
        box: "dx-area-box",
        caption: "dx-area-caption",
        icon: "dx-area-icon",
        field: "dx-area-field",
        fieldContainer: "dx-area-field-container",
        fieldContent: "dx-area-field-content",
        fieldList: "dx-area-fields",
        fieldListHeader: "dx-area-fields-header"
    },
    pivotGrid: {
        dragAction: "dx-pivotgrid-drag-action",
        fieldsContainer: "dx-pivotgrid-fields-container"
    },
    fieldChooser: {
        self: "dx-pivotgridfieldchooser",
        container: "dx-pivotgridfieldchooser-container",
        contextMenu: "dx-pivotgridfieldchooser-context-menu"
    },
    layout: {
        zero: "dx-layout-0",
        second: "dx-layout-2"
    },
    treeView: {
        self: "dx-treeview",
        borderVisible: "dx-treeview-border-visible"
    },
    scrollable: {
        self: "dx-scrollable"
    },
    allFields: "dx-all-fields",
    col: "dx-col",
    headerFilter: "dx-header-filter",
    row: "dx-row",
    widget: "dx-widget"
};
exports.CLASSES = CLASSES;
const ICONS = {
    all: "smalliconslayout",
    column: "columnfield",
    row: "rowfield",
    filter: "filter",
    data: "formula",
    measure: "formula",
    hierarchy: "hierarchy",
    dimension: "detailslayout"
};
exports.ICONS = ICONS;
const SORTABLE_CONST = {
    targets: {
        drag: "drag"
    }
};
exports.SORTABLE_CONST = SORTABLE_CONST;
const SORT_ORDER = {
    descending: "desc",
    ascending: "asc"
};
exports.SORT_ORDER = SORT_ORDER;
