// devextreme/cjs/renovation/ui/scheduler/workspaces/const.js
/**
 * DevExtreme (cjs/renovation/ui/scheduler/workspaces/const.js)
 * Version: 23.2.5
 * Build date: Mon Mar 11 2024
 *
 * Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
exports.DefaultSizes = exports.DATE_TABLE_ROW_CLASS = exports.DATE_TABLE_CELL_CLASS = exports.ALL_DAY_ROW_CLASS = exports.ALL_DAY_PANEL_CELL_CLASS = void 0;
const DefaultSizes = {
    allDayPanelHeight: 25
};
exports.DefaultSizes = DefaultSizes;
const DATE_TABLE_CELL_CLASS = "dx-scheduler-date-table-cell";
exports.DATE_TABLE_CELL_CLASS = DATE_TABLE_CELL_CLASS;
const ALL_DAY_PANEL_CELL_CLASS = "dx-scheduler-all-day-table-cell";
exports.ALL_DAY_PANEL_CELL_CLASS = ALL_DAY_PANEL_CELL_CLASS;
const DATE_TABLE_ROW_CLASS = "dx-scheduler-date-table-row";
exports.DATE_TABLE_ROW_CLASS = DATE_TABLE_ROW_CLASS;
const ALL_DAY_ROW_CLASS = "dx-scheduler-all-day-table-row";
exports.ALL_DAY_ROW_CLASS = ALL_DAY_ROW_CLASS;
