// devextreme/cjs/viz/vector_map/control_bar/utils.js
/**
 * DevExtreme (cjs/viz/vector_map/control_bar/utils.js)
 * Version: 23.2.5
 * Build date: Mon Mar 11 2024
 *
 * Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
exports.toggleDisplay = exports.createVisibilityGroup = exports.createTracker = void 0;
const createTracker = (renderer, root) => renderer.g().attr({
    stroke: "none",
    "stroke-width": 0,
    fill: "#000000",
    opacity: 1e-4
}).css({
    cursor: "pointer"
}).append(root);
exports.createTracker = createTracker;
const createVisibilityGroup = function(renderer, root) {
    let className = arguments.length > 2 && void 0 !== arguments[2] ? arguments[2] : "";
    return renderer.g().attr({
        class: className
    }).append(root)
};
exports.createVisibilityGroup = createVisibilityGroup;
const toggleDisplay = (blocks, isVisible) => {
    const style = isVisible ? {
        display: "block"
    } : {
        display: "none"
    };
    blocks.map(item => item.css(style))
};
exports.toggleDisplay = toggleDisplay;
