// devextreme/cjs/__internal/grids/grid_core/columns_controller/const.js
/**
 * DevExtreme (cjs/__internal/grids/grid_core/columns_controller/const.js)
 * Version: 23.2.5
 * Build date: Mon Mar 11 2024
 *
 * Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.USER_STATE_FIELD_NAMES_15_1 = exports.USER_STATE_FIELD_NAMES = exports.MAX_SAFE_INTEGER = exports.IGNORE_COLUMN_OPTION_NAMES = exports.GROUP_LOCATION = exports.GROUP_COMMAND_COLUMN_NAME = exports.DETAIL_COMMAND_COLUMN_NAME = exports.DEFAULT_COLUMN_OPTIONS = exports.DATATYPE_OPERATIONS = exports.COMMAND_EXPAND_CLASS = exports.COLUMN_OPTION_REGEXP = exports.COLUMN_INDEX_OPTIONS = exports.COLUMN_CHOOSER_LOCATION = void 0;
const USER_STATE_FIELD_NAMES_15_1 = ["filterValues", "filterType", "fixed", "fixedPosition"];
exports.USER_STATE_FIELD_NAMES_15_1 = USER_STATE_FIELD_NAMES_15_1;
const USER_STATE_FIELD_NAMES = ["visibleIndex", "dataField", "name", "dataType", "width", "visible", "sortOrder", "lastSortOrder", "sortIndex", "groupIndex", "filterValue", "bufferedFilterValue", "selectedFilterOperation", "bufferedSelectedFilterOperation", "added"].concat(USER_STATE_FIELD_NAMES_15_1);
exports.USER_STATE_FIELD_NAMES = USER_STATE_FIELD_NAMES;
const IGNORE_COLUMN_OPTION_NAMES = {
    visibleWidth: true,
    bestFitWidth: true,
    bufferedFilterValue: true
};
exports.IGNORE_COLUMN_OPTION_NAMES = IGNORE_COLUMN_OPTION_NAMES;
const COMMAND_EXPAND_CLASS = "dx-command-expand";
exports.COMMAND_EXPAND_CLASS = "dx-command-expand";
const MAX_SAFE_INTEGER = Number.MAX_SAFE_INTEGER || 9007199254740991;
exports.MAX_SAFE_INTEGER = MAX_SAFE_INTEGER;
const GROUP_COMMAND_COLUMN_NAME = "groupExpand";
exports.GROUP_COMMAND_COLUMN_NAME = "groupExpand";
const DETAIL_COMMAND_COLUMN_NAME = "detailExpand";
exports.DETAIL_COMMAND_COLUMN_NAME = "detailExpand";
const COLUMN_OPTION_REGEXP = /columns\[(\d+)\]\.?/gi;
exports.COLUMN_OPTION_REGEXP = COLUMN_OPTION_REGEXP;
const DEFAULT_COLUMN_OPTIONS = {
    visible: true,
    showInColumnChooser: true
};
exports.DEFAULT_COLUMN_OPTIONS = DEFAULT_COLUMN_OPTIONS;
const DATATYPE_OPERATIONS = {
    number: ["=", "<>", "<", ">", "<=", ">=", "between"],
    string: ["contains", "notcontains", "startswith", "endswith", "=", "<>"],
    date: ["=", "<>", "<", ">", "<=", ">=", "between"],
    datetime: ["=", "<>", "<", ">", "<=", ">=", "between"]
};
exports.DATATYPE_OPERATIONS = DATATYPE_OPERATIONS;
const COLUMN_INDEX_OPTIONS = {
    visibleIndex: true,
    groupIndex: true,
    grouped: true,
    sortIndex: true,
    sortOrder: true
};
exports.COLUMN_INDEX_OPTIONS = COLUMN_INDEX_OPTIONS;
const GROUP_LOCATION = "group";
exports.GROUP_LOCATION = "group";
const COLUMN_CHOOSER_LOCATION = "columnChooser";
exports.COLUMN_CHOOSER_LOCATION = "columnChooser";
