// devextreme/cjs/renovation/ui/scheduler/appointment_edit_form/edit_form/layout_items/const.js
/**
 * DevExtreme (cjs/renovation/ui/scheduler/appointment_edit_form/edit_form/layout_items/const.js)
 * Version: 23.2.5
 * Build date: Mon Mar 11 2024
 *
 * Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
exports.ItemLabels = void 0;
var _message = _interopRequireDefault(require("../../../../../../localization/message"));

function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    }
}
const ItemLabels = {
    subject: _message.default.format("dxScheduler-editorLabelTitle"),
    startDate: _message.default.format("dxScheduler-editorLabelStartDate"),
    endDate: _message.default.format("dxScheduler-editorLabelEndDate"),
    allDay: _message.default.format("dxScheduler-allDay"),
    repeat: _message.default.format("dxScheduler-editorLabelRecurrence"),
    description: _message.default.format("dxScheduler-editorLabelDescription")
};
exports.ItemLabels = ItemLabels;
