// devextreme-react/esm/gantt.js
/*!
 * devextreme-react
 * Version: 23.2.5
 * Build date: Mon Mar 11 2024
 *
 * Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-react
 */

"use client";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import dxGantt from "devextreme/ui/gantt";
import * as PropTypes from "prop-types";
import { Component as BaseComponent } from "./core/component";
import NestedOption from "./core/nested-option";
var Gantt = /** @class */ (function (_super) {
    __extends(Gantt, _super);
    function Gantt() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this._WidgetClass = dxGantt;
        _this.independentEvents = ["onContentReady", "onContextMenuPreparing", "onCustomCommand", "onDependencyDeleted", "onDependencyDeleting", "onDependencyInserted", "onDependencyInserting", "onDisposing", "onInitialized", "onResourceAssigned", "onResourceAssigning", "onResourceDeleted", "onResourceDeleting", "onResourceInserted", "onResourceInserting", "onResourceManagerDialogShowing", "onResourceUnassigned", "onResourceUnassigning", "onScaleCellPrepared", "onTaskClick", "onTaskDblClick", "onTaskDeleted", "onTaskDeleting", "onTaskEditDialogShowing", "onTaskInserted", "onTaskInserting", "onTaskMoving", "onTaskUpdated", "onTaskUpdating"];
        _this._expectedChildren = {
            column: { optionName: "columns", isCollectionItem: true },
            contextMenu: { optionName: "contextMenu", isCollectionItem: false },
            dependencies: { optionName: "dependencies", isCollectionItem: false },
            editing: { optionName: "editing", isCollectionItem: false },
            filterRow: { optionName: "filterRow", isCollectionItem: false },
            ganttHeaderFilter: { optionName: "headerFilter", isCollectionItem: false },
            headerFilter: { optionName: "headerFilter", isCollectionItem: false },
            resourceAssignments: { optionName: "resourceAssignments", isCollectionItem: false },
            resources: { optionName: "resources", isCollectionItem: false },
            scaleTypeRange: { optionName: "scaleTypeRange", isCollectionItem: false },
            sorting: { optionName: "sorting", isCollectionItem: false },
            stripLine: { optionName: "stripLines", isCollectionItem: true },
            tasks: { optionName: "tasks", isCollectionItem: false },
            toolbar: { optionName: "toolbar", isCollectionItem: false },
            validation: { optionName: "validation", isCollectionItem: false }
        };
        _this._templateProps = [{
                tmplOption: "taskContentTemplate",
                render: "taskContentRender",
                component: "taskContentComponent",
                keyFn: "taskContentKeyFn"
            }, {
                tmplOption: "taskProgressTooltipContentTemplate",
                render: "taskProgressTooltipContentRender",
                component: "taskProgressTooltipContentComponent",
                keyFn: "taskProgressTooltipContentKeyFn"
            }, {
                tmplOption: "taskTimeTooltipContentTemplate",
                render: "taskTimeTooltipContentRender",
                component: "taskTimeTooltipContentComponent",
                keyFn: "taskTimeTooltipContentKeyFn"
            }, {
                tmplOption: "taskTooltipContentTemplate",
                render: "taskTooltipContentRender",
                component: "taskTooltipContentComponent",
                keyFn: "taskTooltipContentKeyFn"
            }];
        return _this;
    }
    Object.defineProperty(Gantt.prototype, "instance", {
        get: function () {
            return this._instance;
        },
        enumerable: false,
        configurable: true
    });
    return Gantt;
}(BaseComponent));
Gantt.propTypes = {
    accessKey: PropTypes.string,
    activeStateEnabled: PropTypes.bool,
    allowSelection: PropTypes.bool,
    columns: PropTypes.array,
    contextMenu: PropTypes.object,
    dependencies: PropTypes.object,
    disabled: PropTypes.bool,
    editing: PropTypes.object,
    elementAttr: PropTypes.object,
    endDateRange: PropTypes.instanceOf(Date),
    filterRow: PropTypes.object,
    firstDayOfWeek: PropTypes.oneOfType([
        PropTypes.number,
        PropTypes.oneOf([
            0,
            1,
            2,
            3,
            4,
            5,
            6
        ])
    ]),
    focusStateEnabled: PropTypes.bool,
    headerFilter: PropTypes.object,
    height: PropTypes.oneOfType([
        PropTypes.func,
        PropTypes.number,
        PropTypes.string
    ]),
    hint: PropTypes.string,
    hoverStateEnabled: PropTypes.bool,
    onContentReady: PropTypes.func,
    onContextMenuPreparing: PropTypes.func,
    onCustomCommand: PropTypes.func,
    onDependencyDeleted: PropTypes.func,
    onDependencyDeleting: PropTypes.func,
    onDependencyInserted: PropTypes.func,
    onDependencyInserting: PropTypes.func,
    onDisposing: PropTypes.func,
    onInitialized: PropTypes.func,
    onOptionChanged: PropTypes.func,
    onResourceAssigned: PropTypes.func,
    onResourceAssigning: PropTypes.func,
    onResourceDeleted: PropTypes.func,
    onResourceDeleting: PropTypes.func,
    onResourceInserted: PropTypes.func,
    onResourceInserting: PropTypes.func,
    onResourceManagerDialogShowing: PropTypes.func,
    onResourceUnassigned: PropTypes.func,
    onResourceUnassigning: PropTypes.func,
    onScaleCellPrepared: PropTypes.func,
    onSelectionChanged: PropTypes.func,
    onTaskClick: PropTypes.func,
    onTaskDblClick: PropTypes.func,
    onTaskDeleted: PropTypes.func,
    onTaskDeleting: PropTypes.func,
    onTaskEditDialogShowing: PropTypes.func,
    onTaskInserted: PropTypes.func,
    onTaskInserting: PropTypes.func,
    onTaskMoving: PropTypes.func,
    onTaskUpdated: PropTypes.func,
    onTaskUpdating: PropTypes.func,
    resourceAssignments: PropTypes.object,
    resources: PropTypes.object,
    scaleType: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.oneOf([
            "auto",
            "minutes",
            "hours",
            "sixHours",
            "days",
            "weeks",
            "months",
            "quarters",
            "years"
        ])
    ]),
    scaleTypeRange: PropTypes.object,
    showDependencies: PropTypes.bool,
    showResources: PropTypes.bool,
    showRowLines: PropTypes.bool,
    sorting: PropTypes.object,
    startDateRange: PropTypes.instanceOf(Date),
    stripLines: PropTypes.array,
    tabIndex: PropTypes.number,
    taskListWidth: PropTypes.number,
    tasks: PropTypes.object,
    taskTitlePosition: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.oneOf([
            "inside",
            "outside",
            "none"
        ])
    ]),
    toolbar: PropTypes.object,
    validation: PropTypes.object,
    visible: PropTypes.bool,
    width: PropTypes.oneOfType([
        PropTypes.func,
        PropTypes.number,
        PropTypes.string
    ])
};
var Column = /** @class */ (function (_super) {
    __extends(Column, _super);
    function Column() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Column.OptionName = "columns";
    Column.IsCollectionItem = true;
    Column.DefaultsProps = {
        defaultFilterValue: "filterValue",
        defaultFilterValues: "filterValues",
        defaultSelectedFilterOperation: "selectedFilterOperation",
        defaultSortIndex: "sortIndex",
        defaultSortOrder: "sortOrder",
        defaultVisible: "visible",
        defaultVisibleIndex: "visibleIndex"
    };
    Column.ExpectedChildren = {
        columnHeaderFilter: { optionName: "headerFilter", isCollectionItem: false },
        format: { optionName: "format", isCollectionItem: false },
        headerFilter: { optionName: "headerFilter", isCollectionItem: false }
    };
    Column.TemplateProps = [{
            tmplOption: "cellTemplate",
            render: "cellRender",
            component: "cellComponent",
            keyFn: "cellKeyFn"
        }, {
            tmplOption: "headerCellTemplate",
            render: "headerCellRender",
            component: "headerCellComponent",
            keyFn: "headerCellKeyFn"
        }];
    return Column;
}(NestedOption));
var ColumnHeaderFilter = /** @class */ (function (_super) {
    __extends(ColumnHeaderFilter, _super);
    function ColumnHeaderFilter() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ColumnHeaderFilter.OptionName = "headerFilter";
    ColumnHeaderFilter.ExpectedChildren = {
        columnHeaderFilterSearch: { optionName: "search", isCollectionItem: false },
        search: { optionName: "search", isCollectionItem: false }
    };
    return ColumnHeaderFilter;
}(NestedOption));
var ColumnHeaderFilterSearch = /** @class */ (function (_super) {
    __extends(ColumnHeaderFilterSearch, _super);
    function ColumnHeaderFilterSearch() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ColumnHeaderFilterSearch.OptionName = "search";
    return ColumnHeaderFilterSearch;
}(NestedOption));
var ContextMenu = /** @class */ (function (_super) {
    __extends(ContextMenu, _super);
    function ContextMenu() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ContextMenu.OptionName = "contextMenu";
    ContextMenu.ExpectedChildren = {
        contextMenuItem: { optionName: "items", isCollectionItem: true },
        item: { optionName: "items", isCollectionItem: true }
    };
    return ContextMenu;
}(NestedOption));
var ContextMenuItem = /** @class */ (function (_super) {
    __extends(ContextMenuItem, _super);
    function ContextMenuItem() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ContextMenuItem.OptionName = "items";
    ContextMenuItem.IsCollectionItem = true;
    ContextMenuItem.TemplateProps = [{
            tmplOption: "template",
            render: "render",
            component: "component",
            keyFn: "keyFn"
        }];
    return ContextMenuItem;
}(NestedOption));
var Dependencies = /** @class */ (function (_super) {
    __extends(Dependencies, _super);
    function Dependencies() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Dependencies.OptionName = "dependencies";
    return Dependencies;
}(NestedOption));
var Editing = /** @class */ (function (_super) {
    __extends(Editing, _super);
    function Editing() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Editing.OptionName = "editing";
    return Editing;
}(NestedOption));
var FilterRow = /** @class */ (function (_super) {
    __extends(FilterRow, _super);
    function FilterRow() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    FilterRow.OptionName = "filterRow";
    FilterRow.ExpectedChildren = {
        operationDescriptions: { optionName: "operationDescriptions", isCollectionItem: false }
    };
    return FilterRow;
}(NestedOption));
var Format = /** @class */ (function (_super) {
    __extends(Format, _super);
    function Format() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Format.OptionName = "format";
    return Format;
}(NestedOption));
var GanttHeaderFilter = /** @class */ (function (_super) {
    __extends(GanttHeaderFilter, _super);
    function GanttHeaderFilter() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    GanttHeaderFilter.OptionName = "headerFilter";
    GanttHeaderFilter.ExpectedChildren = {
        ganttHeaderFilterSearch: { optionName: "search", isCollectionItem: false },
        search: { optionName: "search", isCollectionItem: false },
        texts: { optionName: "texts", isCollectionItem: false }
    };
    return GanttHeaderFilter;
}(NestedOption));
var GanttHeaderFilterSearch = /** @class */ (function (_super) {
    __extends(GanttHeaderFilterSearch, _super);
    function GanttHeaderFilterSearch() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    GanttHeaderFilterSearch.OptionName = "search";
    return GanttHeaderFilterSearch;
}(NestedOption));
var HeaderFilter = /** @class */ (function (_super) {
    __extends(HeaderFilter, _super);
    function HeaderFilter() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    HeaderFilter.OptionName = "headerFilter";
    return HeaderFilter;
}(NestedOption));
var Item = /** @class */ (function (_super) {
    __extends(Item, _super);
    function Item() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Item.OptionName = "items";
    Item.IsCollectionItem = true;
    Item.TemplateProps = [{
            tmplOption: "template",
            render: "render",
            component: "component",
            keyFn: "keyFn"
        }, {
            tmplOption: "menuItemTemplate",
            render: "menuItemRender",
            component: "menuItemComponent",
            keyFn: "menuItemKeyFn"
        }];
    return Item;
}(NestedOption));
var OperationDescriptions = /** @class */ (function (_super) {
    __extends(OperationDescriptions, _super);
    function OperationDescriptions() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    OperationDescriptions.OptionName = "operationDescriptions";
    return OperationDescriptions;
}(NestedOption));
var ResourceAssignments = /** @class */ (function (_super) {
    __extends(ResourceAssignments, _super);
    function ResourceAssignments() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ResourceAssignments.OptionName = "resourceAssignments";
    return ResourceAssignments;
}(NestedOption));
var Resources = /** @class */ (function (_super) {
    __extends(Resources, _super);
    function Resources() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Resources.OptionName = "resources";
    return Resources;
}(NestedOption));
var ScaleTypeRange = /** @class */ (function (_super) {
    __extends(ScaleTypeRange, _super);
    function ScaleTypeRange() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ScaleTypeRange.OptionName = "scaleTypeRange";
    return ScaleTypeRange;
}(NestedOption));
var Search = /** @class */ (function (_super) {
    __extends(Search, _super);
    function Search() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Search.OptionName = "search";
    return Search;
}(NestedOption));
var Sorting = /** @class */ (function (_super) {
    __extends(Sorting, _super);
    function Sorting() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Sorting.OptionName = "sorting";
    return Sorting;
}(NestedOption));
var StripLine = /** @class */ (function (_super) {
    __extends(StripLine, _super);
    function StripLine() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    StripLine.OptionName = "stripLines";
    StripLine.IsCollectionItem = true;
    return StripLine;
}(NestedOption));
var Tasks = /** @class */ (function (_super) {
    __extends(Tasks, _super);
    function Tasks() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Tasks.OptionName = "tasks";
    return Tasks;
}(NestedOption));
var Texts = /** @class */ (function (_super) {
    __extends(Texts, _super);
    function Texts() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Texts.OptionName = "texts";
    return Texts;
}(NestedOption));
var Toolbar = /** @class */ (function (_super) {
    __extends(Toolbar, _super);
    function Toolbar() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Toolbar.OptionName = "toolbar";
    Toolbar.ExpectedChildren = {
        item: { optionName: "items", isCollectionItem: true },
        toolbarItem: { optionName: "items", isCollectionItem: true }
    };
    return Toolbar;
}(NestedOption));
var ToolbarItem = /** @class */ (function (_super) {
    __extends(ToolbarItem, _super);
    function ToolbarItem() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ToolbarItem.OptionName = "items";
    ToolbarItem.IsCollectionItem = true;
    ToolbarItem.TemplateProps = [{
            tmplOption: "menuItemTemplate",
            render: "menuItemRender",
            component: "menuItemComponent",
            keyFn: "menuItemKeyFn"
        }, {
            tmplOption: "template",
            render: "render",
            component: "component",
            keyFn: "keyFn"
        }];
    return ToolbarItem;
}(NestedOption));
var Validation = /** @class */ (function (_super) {
    __extends(Validation, _super);
    function Validation() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Validation.OptionName = "validation";
    return Validation;
}(NestedOption));
export default Gantt;
export { Gantt, Column, ColumnHeaderFilter, ColumnHeaderFilterSearch, ContextMenu, ContextMenuItem, Dependencies, Editing, FilterRow, Format, GanttHeaderFilter, GanttHeaderFilterSearch, HeaderFilter, Item, OperationDescriptions, ResourceAssignments, Resources, ScaleTypeRange, Search, Sorting, StripLine, Tasks, Texts, Toolbar, ToolbarItem, Validation };
