// devextreme-react/esm/scroll-view.js
/*!
 * devextreme-react
 * Version: 23.2.5
 * Build date: Mon Mar 11 2024
 *
 * Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-react
 */

"use client";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import dxScrollView from "devextreme/ui/scroll_view";
import * as PropTypes from "prop-types";
import { Component as BaseComponent } from "./core/component";
var ScrollView = /** @class */ (function (_super) {
    __extends(ScrollView, _super);
    function ScrollView() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this._WidgetClass = dxScrollView;
        _this.independentEvents = ["onDisposing", "onInitialized", "onPullDown", "onReachBottom", "onScroll", "onUpdated"];
        return _this;
    }
    Object.defineProperty(ScrollView.prototype, "instance", {
        get: function () {
            return this._instance;
        },
        enumerable: false,
        configurable: true
    });
    return ScrollView;
}(BaseComponent));
ScrollView.propTypes = {
    bounceEnabled: PropTypes.bool,
    direction: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.oneOf([
            "both",
            "horizontal",
            "vertical"
        ])
    ]),
    disabled: PropTypes.bool,
    elementAttr: PropTypes.object,
    height: PropTypes.oneOfType([
        PropTypes.func,
        PropTypes.number,
        PropTypes.string
    ]),
    onDisposing: PropTypes.func,
    onInitialized: PropTypes.func,
    onOptionChanged: PropTypes.func,
    onPullDown: PropTypes.func,
    onReachBottom: PropTypes.func,
    onScroll: PropTypes.func,
    onUpdated: PropTypes.func,
    pulledDownText: PropTypes.string,
    pullingDownText: PropTypes.string,
    reachBottomText: PropTypes.string,
    refreshingText: PropTypes.string,
    rtlEnabled: PropTypes.bool,
    scrollByContent: PropTypes.bool,
    scrollByThumb: PropTypes.bool,
    showScrollbar: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.oneOf([
            "onScroll",
            "onHover",
            "always",
            "never"
        ])
    ]),
    useNative: PropTypes.bool,
    width: PropTypes.oneOfType([
        PropTypes.func,
        PropTypes.number,
        PropTypes.string
    ])
};
export default ScrollView;
export { ScrollView };
